mobify/progressive-web-scaffold
# Use cache manifest

**JIRA**: [WEB-726](https://mobify.atlassian.net/browse/WEB-726)
 **Linked PRs**: https://github.com/mobify/progressive-web-sdk/pull/128

## Changes
- Update `.gitignore` to ignore `tmp/` directory
- Adds required, user-editable `config/cache-hash-config.json` configuration file as well as some sane defaults (for the scaffold, anyway)
- Updates loader and main webpack configuration to provide an alias for `cacheHashManifest`
- Update `dev` and `build` npm scripts to call cache manifest binary before starting dev server or building app files, respectively

## Todos
- [x] Add a high-level description of your changes to the CHANGELOG.md.
- [x] Investigate whether it's worthwhile to remove `loader.js` dependencies on `progressive-web-sdk/asset-utils` since it will now include the hash manifest (and possibly bloat file size)

## How to test-drive this PR
- Check out [progressive-web-sdk#use-cache-manifest](https://github.com/mobify/progressive-web-sdk/pull/128) and run `npm link`
- Check out this branch
- run `npm link progressive-web-sdk`
- run `npm run dev`

### Verify:
- `tmp/cache-hash-manifest.json` was created and lists `static/js/jquery.min.js`
- `tmp/loader-cache-hash-manifest.json` was created and has an empty `hashes` object and a buildDate string
- project builds and can be previewed against `http://www.merlinspotions.com`
- rendered HTML applies cachebreakers to `jquery.min.js`, `main.js` and `main.css` that match manifest `buildDate` entry
- modifying `hashLength` in `app/config/cache-hash-config.json` modifies generated hashes

### Bonus:
- Add `import {getAssetUrl} from 'progressive-web-sdk/src/asset-utils'` to a container file
- Add a new image to the project somewhere in `static/`
- Reference that image using `getAssetUrl` e.g. `<img src={getAssetUrl('static/img/hello.png')} />`
- Run `npm run dev`
- Rendered HTML shows the image you imported into the project, with a md5 hash applied to the src path that matches what's in the cache manifest
